src/common/src/array/data_chunk.rs

```diff
@@ -57,7 +57,24 @@ impl DataChunkBuilder {
     }
 
     pub fn build(self) -> DataChunk {
-        DataChunk::new(self.columns, self.visibility)
+        let vis = if let Some(bitmap) = self.visibility {
+            // with visibility bitmap
+            for column in &self.columns {
+                assert_eq!(bitmap.len(), column.array_ref().len())
+            }
+            Vis::Bitmap(bitmap)
+        } else if !self.columns.is_empty() {
+            // without visibility bitmap
+            let card = self.columns.first().unwrap().array_ref().len();
+            for column in self.columns.iter().skip(1) {
+                assert_eq!(card, column.array_ref().len())
+            }
+            Vis::Compact(card)
+        } else {
+            // no data (dummy)
+            Vis::Compact(0)
+        };
```

Comment:
Problematic logic moved from previous `DataChunk::new` to keep compatible with current `DataChunkBuilder` callers. This will be removed in refactor part 3/3.